### DISCUSSION TITLE
No convergence with an Mfront law in code_aster

### DISCUSSION THREAD

**Discussion by Max-EDFLMT** [thelfer/tfel#182] [Q&A] [answered]
Hi,

I'm trying to use an MFront law in a code_aster model. it is adapted from the one sent in #130 (I added Voce hardening law to it).

It's working as expected in Mtest and on a small domain with periodic boundary conditions in code_aster.
When I try on my real structure, code_aster never success in integrating the law. It works with a native aster law on the same model or with Mfront FCC crystal plasticity law.

I tried the keyword @AsterGenerateMTestFileOnFailure but I don't know where I'm supposed to find the files in the computation repository.

Any hints on how to debug this ?

Thanks in advance,

Maxime

[OrthotropicHosford_Voce.txt](https://github.com/thelfer/tfel/files/8642222/OrthotropicHosford_Voce.txt)

**Comment by thelfer** [comment, marked as answer]
Hi @Max-EDFLMT.

Good to see that you made progresses. Very interested to see what will be the final result

First of all, have you read the last part of [this tutorial](https://thelfer.github.io/tfel/web/hosford.html) ?
You probably shall use the Jacobi eigen solver and either Algorithm 3 or Algorithm 4.

Concerning `code_aster` and the `MTest` file generation, I don't remember the details, but the basic idea is to put the call the `STAT_NON_LINE` in  `try/except` bloc and do something like this:

```python
try:
    STAT_NON_LINE(...)
except:
    os.system("cp *.mtest <insert_directory_where_to_copy_the_mtest_files>")
```

**Reply by Max-EDFLMT** [reply]
Hi Thomas,

I did a lot of tests following the tutorial and actually, I have also tried your Barlat implementation with coefficients that should normally give the same behavior than the file above. I had the same law integration issue for any power value (even for low ones that do not require the precise developments discussed in the tutorial, do they ?)

I will try the mtest generation on failure. The name before ".mtest" of the file is supposed to match something declared before ? Or how is it supposed to know what to write otherwise ?

Thanks !

Maxime

**Reply by thelfer** [reply]
> I did a lot of tests following the tutorial and actually, I have also tried your Barlat implementation with coefficients that should normally give the same behavior than the file above. I had the same law integration issue for any power value (even for low ones that do not require the precise developments discussed in the tutorial, do they ?)

Strange. The Barlat implementation was used for metal forming in `Abaqus` and did not lead to any problem. You may also want to add `@AsterErrorReport true;` in your MFront file. It may help.

> The name before ".mtest" of the file is supposed to match something declared before ?

`*.mtest` (do not forget the star) is a regular expression, meaning all the files in the current directory with extension `.mtest`. Their names will depend on the name of the behaviour.

**Reply by Max-EDFLMT** [reply]
> Strange. The Barlat implementation was used for metal forming in Abaqus and did not lead to any problem. You may also want to add @AsterErrorReport true; in your MFront file. It may help.

I used the standard elastoviscoplasticity brick with a Voce law for this. It was working with mtest but again failing with aster. I will try the recommended keyword !

> *.mtest (do not forget the star) is a regular expression, meaning all the files in the current directory with extension .mtest. Their names will depend on the name of the behaviour.

Of course, I did not see the "cp", it's being late ... Nonetheless, I checked the tmp folder when launching a computation  failing for multiple subdivision of the time step. I did not see any .mtest file so I wonder where it is hidden if there is one. I'll let you know after giving it a try.

**Reply by Max-EDFLMT** [reply]
So I added --@AsterErrorReport=true and --@AsterGenerateMTestFileOnFailure=true.
With the try: ... except: ... I had no *.mtest file to copy as I first expected, looking in the tmp folder during the computation.
For the error report, I don't think I had anything in the ouptut file. Is it supposed to be read elsewhere ?

**Reply by thelfer** [reply]
@Max-EDFLMT I would at least expect an error message telling what went wrong in the terminal, if not in the log file...
Likewise, the fact that you don't have any mtest file generated is quite strange. Are you sure to link the correct library ?

If you don't mind, let us debug further.

1. Start to generate the `C++` sources from `MFront`
    ```bash
   $ mfront --interface=aster OrthotropicHosford_Voce.mfront
    ```
2. Open the generated sources:
    ```bash
   $ src/asterOrthotropicHosfordVoce.cxx
    ```
3. Modify the  OrthotropicHosfordVoce function as follows:
    ```C++
   MFRONT_SHAREDOBJ void
   OrthotropicHosfordVoce(aster::AsterReal *const STRESS,aster::AsterReal *const STATEV,aster::AsterReal *const DDSOE,const aster::AsterReal *const STRAN,const aster::AsterReal *const DSTRAN,const aster::AsterReal *const DTIME,const aster::AsterReal *const TEMP,const aster::AsterReal *const DTEMP,const aster::AsterReal *const PREDEF,const aster::AsterReal *const DPRED,const aster::AsterInt  *const NTENS,const aster::AsterInt  *const NSTATV,const aster::AsterReal *const PROPS,const aster::AsterInt  *const NPROPS,const aster::AsterReal *const DROT,aster::AsterReal *const PNEWDT,const aster::AsterInt *const NUMMOD)
   {
   std::cout << "In OrthotropicHosfordVoce" << std::endl;
    ```
4. Generate the shared library
    ```bash
   $ mfront --obuild
    ```
5. Indicate to `code_aster` the exact path of the library generated (See <https://thelfer.github.io/tfel/web/aster.html> for details).
6. Check that the previous message is printed
7. Once Step 6 is ok, remove the previous message and modify the test next to it as follows:
    ```C++
    if(aster::AsterInterface<tfel::material::OrthotropicHosfordVoce>::exe(msg,...)!=0){
      std::cout << "Something went wrong: " << msg << std::endl;
    ```
8. Regenerate the shared library
    ```bash
   $ mfront --obuild
    ```
9. Re-run the test.

**Reply by Max-EDFLMT** [reply]
I put everything in my aster folder so there is a single mfront file and src directory is in this folder. Thus, I'm sure that I take the relevant library.
I already fail the first test. I have no message "In OrthotropicHosfordVoce" printed in the out or msg file given by aster.

**Reply by thelfer** [reply]
@Max-EDFLMT To be sure, if you just remove the `MFront` shared library and run `code_aster`, does it fail by stating that the shared library can't be loaded ?

**Reply by Max-EDFLMT** [reply]
Yes, it is ! "<INFO> fichier/répertoire inexistant"

**Reply by thelfer** [reply]
Strange, would you try again by using `std::cerr` rather than `std::cout` ?
Are you don't recompile the library at the beginning of your script ?

**Reply by Max-EDFLMT** [reply]
Ok, I have figured out an issue. I have two mfront behaviours in my model. The second one is loaded from an other distant folder but it looks like only this one was used despite the fact that some of my elements are attributed to the OrthotropicHosfordVoce behaviour. That may explain why in my other "simple" model, it was working ... Is this buggy in aster ? This maybe beyond your purview ...

**Reply by Max-EDFLMT** [reply]
It's becoming clearer I guess. I've followed [https://thelfer.github.io/tfel/web/aster.html]() for assigning two different behaviours.
The issue is that between 

> MAT =DEFI_MATERIAU( MFRONT=_F( LISTE_COEF = (C1,C2,....)),)

and 

> COMPORTEMENT=_F (RELATION = 'MFRONT'......), 

you assign a behaviours to elements using 

> AFFE_MATERIAU(MAILLAGE = MESH, AFFE = (_F(GROUP_MA='GR' MATER = MAT, ),

But you never say which file you want to use in DEFI_MATERIAU when doing :

> MAT1 = DEFI_MATERIAU(ELAS=_F(E=E,NU=NU),MFRONT=_F(LISTE_COEF=..... );
> AFFE_MATERIAU(MAILLAGE = MESH, AFFE = (_F(GROUP_MA='GR1' MATER = MAT1, ),
> MAT2 = DEFI_MATERIAU(ELAS=_F(E=E,NU=NU),MFRONT=_F(LISTE_COEF=..... );
> AFFE_MATERIAU(MAILLAGE = MESH, AFFE = (_F(GROUP_MA='GR2' MATER = MAT2, ),

So, my guess was that it was in the same order than when declaring : 

> COMPORTEMENT = (_F(ALGO_CPLAN='ANALYTIQUE',
>                                            RELATION='MFRONT',
>                                            NOM_ROUTINE='astermat1',
>                                            UNITE_LIBRAIRIE=40,
>                                           ),
>                                         _F(ALGO_CPLAN='ANALYTIQUE',
>                                            RELATION='MFRONT',
>                                            NOM_ROUTINE='astermat2',
>                                            UNITE_LIBRAIRIE=41,
>                                           ),                                   
>                                    ),

But it's obviously not working ! Am I doing it wrong ?

**Reply by thelfer** [reply]
@Max-EDFLMT Not sure to have understood everything, but it seems awful at first sight, does it ?
In the end, are you able to make a computation with two distinct MFront materials ?

**Reply by Max-EDFLMT** [reply]
It is !
In the end, I don't think I can, no. And I also think that I misunderstood some of my last models because of the same issue ...

In order to clarify things, the idea is that you never say which mfront behaviour you want when assigning a material. I checked code_aster keywords, there is no such things like that.

When creating a behaviour we assign it to an Mfront handle, passing a list of coefficients to the function but we never say which Mfront file we want to use with this coefficients.
Then, we assign this behaviour to some elements.
Finally the behaviours are declared in STAT_NON_LINE but there is no mention to the elements.

At last, we don't know what was assigned and my guess is that it is only the behaviour declared because in the tests you asked me to conduct I had no feedback until puting the incriminated behaviour at the end.

**Reply by thelfer** [reply]
@Max-EDFLMT, I admit that I am quite disapointed. I asked my favourite `code_aster` developers for help. Something needs clarification. Maybe you shall also post this on the `code_aster` forum.

**Reply by Max-EDFLMT** [reply]
Thank you anyway. I will ask on `code_aster` forum this monday !

**Reply by thelfer** [reply]
My job !

**Reply by Max-EDFLMT** [reply]
It turned out that there is a keyword GROUP_MA in COMPORTEMENT = _F(...) to mention the element group you want to assign the previously mentionned DEFI_MATERIAU. I did not know it and thought everything was to be declared before.
Right now, everything seems to work !

**Reply by thelfer** [reply]
Great !

**Reply by rian44s** [reply]
Hi both, 

I am a bit late on this discussion but I have, I think a very similar problem.
I have a simulation with multiple grains running perfectly in code-aster with a meric-cailletaud law, small strain.
I have downloaded MericCailletaudSingleCrystalViscoPlasticity.mfront to compare.
Then the only modification I have made in aster python launcher file are:

- monMateriau = DEFI_MATERIAU( ELAS =_F(E = young,NU = poisson),
         MFRONT =_F( LISTE_COEF = (young,young,young,poisson,poisson,poisson,cissail,cissail,cissail,
         N,K,tau0,Q,B,D,C,H0,H1,H2,H3,H4,H5,H6 )
         
    instead of 
    
- monMateriau = DEFI_MATERIAU(ELAS_ORTH=_F(E_L= young,E_T= young,E_N= young,G_LT= cissail,G_TN= cissail,G_LN= cissail,NU_LT= poisson,NU_TN= poisson, NU_LN= poisson,),
      MONO_VISC1=_F(N=N, K=K,C=C),
      MONO_ISOT1=_F(R_0=Tau0, Q=Q,B=B,H1=H0,H2=H1,H3=H4,H4=H5,H5=H6),
      MONO_CINE1=_F(D=D))

then      

-   DEFI_FICHIER(FICHIER=path_mfront,UNITE=33)
MonBehaviour = CREA_LIB_MFRONT(NOM_COMPOR=mfront_name,UNITE_MFRONT=33,)
     
     instead of 
     

- MonBehaviour = DEFI_COMPOR(MONOCRISTAL =(_F(MATER = monMateriau,ELAS ='ELAS_ORTH',ECOULEMENT ='MONO_VISC1',ECRO_ISOT = 'MONO_ISOT1',ECRO_CINE = 'MONO_CINE1',FAMI_SYST_GLIS = 'OCTAEDRIQUE')))
      
and in Stat_non_lin

- COMPORTEMENT =_F( RELATION = 'MFRONT',
            COMPOR_MFRONT = MonBehaviour,
            DEFORMATION = 'PETIT',
            ITER_INTE_MAXI = 1000,
            RESI_INTE_MAXI =1.0E-12,
            ),
            
 instead of 
           

- COMPORTEMENT=_F(COMPOR=MonBehaviour,DEFORMATION='PETIT',RELATION='MONOCRISTAL'),


I eventually get this error: 
<EXCEPTION> <MECANONLINE9_3>
Arrêt suite à l'échec de l'intégration de la loi de comportement. 

@Max-EDFLMT you said that adding GROUP_MA in comportement helped but I don't understand, the default value is 'all' and link between behavior and domain is already done in 
fieldmat = AFFE_MATERIAU(AFFE=_F(MATER=(monMateriau, ),TOUT='OUI'),MODELE=monModel), isn't it ?

I would be happy to see what worked for you.

Thank you in advance, 
Rian SEGHIR
